modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk.clj

```diff
@@ -58,6 +59,63 @@
                   (.setCredentials (.createScoped creds bigquery-scopes)))]
     (.. bq-bldr build getService)))
 
+;;; +----------------------------------------------------------------------------------------------------------------+
+;;; |                                         Transducing Query Results                                              |
+;;; +----------------------------------------------------------------------------------------------------------------+
+(def ^:private ^:dynamic *page-callback*
+  "Callback to execute when a new page is retrieved, used for testing"
+  (constantly nil))
+
+(defn- values-iterator ^Iterator [^TableResult page]
+  (.iterator (.getValues page)))
+
+(defn- reducible-bigquery-results
+  [^TableResult page cancel-chan]
+  (reify
+    clojure.lang.IReduceInit
+    (reduce [_ rf init]
+      ;; TODO: Once we're confident that the memory/thread leaks in BigQuery are resolved, we can remove some of this
+      ;; logging, and certainly remove the `n` counter.
+      (loop [^TableResult page page
+             it                (values-iterator page)
+             acc               init
+             n                 0]
+        (cond
+          ;; Early exit: If the cancel-chan is provided, close it. This prevents thread leaks in execute-bigquery.
+          (reduced? acc)
+          (do (log/tracef "BigQuery: Early exit from reducer after %d rows" n)
+              (some-> cancel-chan a/close!)
```

Comment:
What about always providing a `cancel-chan`? Instead of `nil` we could pass in a fresh promise channel.